README.md
# Docker Machine speedycloud Driver

This is a plugin for [Docker Machine](https://docs.docker.com/machine/) allowing
to create Docker hosts locally on SpeedyCloud (http://www.speedycloud.cn/)

## Requirements
* OS X 10.9+
* [Docker Machine](https://docs.docker.com/machine/) 0.5.1+ (is bundled to
  [Docker Toolbox](https://www.docker.com/docker-toolbox) 1.9.1+)
* [speedycloud Desktop](http://www.speedycloud.com/products/desktop/) 11.0.0+ **Pro** or
**Business** edition (_Standard edition is not supported!_)

## Installation
Install via Homebrew:

```console
$ brew install docker-machine-speedycloud
```

To install this plugin manually, download the binary `docker-machine-driver-speedycloud`
and  make it available by `$PATH`, for example by putting it to `/usr/local/bin/`:

```console
$ curl -L https://github.com/speedycloud/docker-machine-speedycloud/releases/download/v1.2.2/docker-machine-driver-speedycloud > /usr/local/bin/docker-machine-driver-speedycloud

$ chmod +x /usr/local/bin/docker-machine-driver-speedycloud
```

The latest version of `docker-machine-driver-speedycloud` binary is available on
the ["Releases"](https://github.com/speedycloud/docker-machine-speedycloud/releases) page.

## Usage
Official documentation for Docker Machine [is available here](https://docs.docker.com/machine/).

To create a speedycloud Desktop virtual machine for Docker purposes just run this
command:

```
$ docker-machine create --driver=speedycloud prl-dev
```

Available options:

 - `--speedycloud-boot2docker-url`: The URL of the boot2docker image.
 - `--speedycloud-disk-size`: Size of disk for the host VM (in MB).
 - `--speedycloud-memory`: Size of memory for the host VM (in MB).
 - `--speedycloud-cpu-count`: Number of CPUs to use to create the VM (-1 to use the number of CPUs available).
 - `--speedycloud-no-share`: Disable the sharing of `/Users` directory

The `--speedycloud-boot2docker-url` flag takes a few different forms. By
default, if no value is specified for this flag, Machine will check locally for
a boot2docker ISO. If one is found, that will be used as the ISO for the
created machine. If one is not found, the latest ISO release available on
[boot2docker/boot2docker](https://github.com/boot2docker/boot2docker) will be
downloaded and stored locally for future use. Note that this means you must run
`docker-machine upgrade` deliberately on a machine if you wish to update the "cached"
boot2docker ISO.

This is the default behavior (when `--speedycloud-boot2docker-url=""`), but the
option also supports specifying ISOs by the `http://` and `file://` protocols.

Environment variables and default values:

| CLI option                    | Environment variable        | Default                  |
|-------------------------------|-----------------------------|--------------------------|
| `--speedycloud-boot2docker-url` | `speedycloud_BOOT2DOCKER_URL` | *Latest boot2docker url* |
| `--speedycloud-cpu-count`       | `speedycloud_CPU_COUNT`       | `1`                      |
| `--speedycloud-disk-size`       | `speedycloud_DISK_SIZE`       | `20000`                  |
| `--speedycloud-memory`          | `speedycloud_MEMORY_SIZE`     | `1024`                   |
| `--speedycloud-no-share`        | -                           | `false`                  |

## Development

### Build from Source
If you wish to work on speedycloud Driver for Docker machine, you'll first need
[Go](http://www.golang.org) installed (version 1.7+ is required).
Make sure Go is properly installed, including setting up a [GOPATH](http://golang.org/doc/code.html#GOPATH).

Run these commands to build the plugin binary:

```bash
$ go get -d github.com/speedycloud/docker-machine-speedycloud
$ cd $GOPATH/github.com/speedycloud/docker-machine-speedycloud
$ make build
```

After the build is complete, `bin/docker-machine-driver-speedycloud` binary will
be created. If you want to copy it to the `${GOPATH}/bin/`, run `make install`.

### Acceptance Tests

We use [BATS](https://github.com/sstephenson/bats) for acceptance testing, so,
[install it](https://github.com/sstephenson/bats#installing-bats-from-source) first.

You also need to build the plugin binary by calling `make build`.

Then you can run acceptance tests using this command:

```bash
$ make test-acceptance
```

Acceptance tests will invoke the general `docker-machine` binary available by
`$PATH`. If you want to specify it explicitly, just set `MACHINE_BINARY` env variable:

```bash
$ MACHINE_BINARY=/path/to/docker-machine make test-acceptance
```

## Authors

* Mikhail Zholobov ([@legal90](https://github.com/legal90))
* Rickard von Essen ([@rickard-von-essen](https://github.com/rickard-von-essen))


CHANGELOG.md
# Changelog

## 1.2.2 (January 10, 2017)
- Use Golang 1.7 to build the binary, update dependencies

## 1.2.1 (July 12, 2016)
- Force to connect the attached ISO image
- Check whether the host is connected to Shared network interface [GH-56]
- Update build dependencies [GH-57]

## 1.2.0 (April 15, 2016)
- Enable time sync feature by default [GH-47]
- Use Golang 1.6 to build the binary, update dependencies [GH-46]
- Update Boot2Docker cache in `PreCreateCheck` [GH-48]

## 1.1.1 (December 29, 2015)
- Compatibility with Docker Machine v0.5.5+ [GH-33]
- Add arguments `--version` and `--help` for plugin binary [GH-34]

## 1.1.0 (November 20, 2015)
- Compatibility with Docker Machine v0.5.1+

## 1.0.2 (November 20, 2015)
- Fixed issue with relative path in `MACHINE_STORAGE_PATH` [GH-17]

## 1.0.1 (November 11, 2015)
- Add unofficial support of speedycloud Desktop 10 [GH-10]
- Remove `--nested-virt` and `--pmu-virt` from default VM settings [GH-13]

## 1.0.0 (November 4, 2015)
Initial public release
- Add `--speedycloud-no-share` flag [GH-1]
- Add acceptance tests
- Add build scripts


speedycloud_test.go
package speedycloud

import (
	"testing"

	"github.com/docker/machine/libmachine/drivers"
	"github.com/stretchr/testify/assert"
)

func TestSetConfigFromFlags(t *testing.T) {
	driver := NewDriver("default", "path")

	checkFlags := &drivers.CheckDriverOptions{
		FlagsValues: map[string]interface{}{
			"openstack-auth-url":  "http://url",
			"openstack-username":  "user",
			"openstack-password":  "pwd",
			"openstack-tenant-id": "ID",
			"openstack-flavor-id": "ID",
			"openstack-image-id":  "ID",
		},
		CreateFlags: driver.GetCreateFlags(),
	}

	err := driver.SetConfigFromFlags(checkFlags)

	assert.NoError(t, err)
	assert.Empty(t, checkFlags.InvalidFlags)
}


test/integration/run-bats.sh
#!/bin/bash

set -e

# Wrapper script to run bats tests for "speedycloud" driver.
# Usage: ./run-bats.sh [subtest]

function quiet_run () {
    if [[ "$VERBOSE" == "1" ]]; then
        "$@"
    else
        "$@" &>/dev/null
    fi
}

function cleanup_machines() {
    if [[ $(machine ls -q | wc -l) -ne 0 ]]; then
        #quiet_run machine stop $(machine ls -q)
        quiet_run machine rm -f $(machine ls -q)
    fi
}

function machine() {
    export PATH="$PLUGIN_ROOT"/bin:$PATH
    "$MACHINE_BINARY" "$@"
}

function run_bats() {
    for bats_file in $(find "$1" -name \*.bats); do
        # BATS returns non-zero to indicate the tests have failed, we shouldn't
        # neccesarily bail in this case, so that's the reason for the e toggle.
        set +e
        echo "=> $bats_file"
        bats "$bats_file"
        if [[ $? -ne 0 ]]; then
            EXIT_STATUS=1
        fi
        set -e
        echo
        cleanup_machines
    done
}

# Set this ourselves in case bats call fails
EXIT_STATUS=0
export BATS_FILE="$1"

if [[ -z "$BATS_FILE" ]]; then
    echo "You must specify a bats test to run."
    exit 1
fi

if [[ ! -e "$BATS_FILE" ]]; then
    echo "Requested bats file or directory not found: $BATS_FILE"
    exit 1
fi

# Set defaults if variables are not defined
export MACHINE_BINARY=${MACHINE_BINARY:-"docker-machine"}

export DRIVER="speedycloud"
export BASE_TEST_DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
export PLUGIN_ROOT="$BASE_TEST_DIR/../.."
export NAME="bats-$DRIVER-test"
export MACHINE_STORAGE_PATH="/tmp/machine-bats-test-$DRIVER"
export BATS_LOG="$PLUGIN_ROOT/bats.log"

# This function gets used in the integration tests, so export it.
export -f machine

rm -f "$BATS_LOG"

run_bats "$BATS_FILE"

if [[ -d "$MACHINE_STORAGE_PATH" ]]; then
    rm -r "$MACHINE_STORAGE_PATH"
fi

exit ${EXIT_STATUS}


speedycloud.go
package speedycloud

import (
    "fmt"
    "io/ioutil"
    "net"
    "strings"
    "time"

    "github.com/docker/machine/libmachine/drivers"
    "github.com/docker/machine/libmachine/log"
    "github.com/docker/machine/libmachine/mcnflag"
    "github.com/docker/machine/libmachine/mcnutils"
    "github.com/docker/machine/libmachine/ssh"
    "github.com/docker/machine/libmachine/state"
    "regexp"
)

type Driver struct {
    *drivers.BaseDriver
    ActiveTimeout    int
    SpeedCloudUrl    string
    ApiKey           string
    ApiSecret        string
    AvailabilityZone string
    MachineId        string
    KeyPairName      string
    NetworkName      string
    UserData         string
    PrivateKeyFile   string
    CpuNumber        int
    Memory           int
    DiskType         string
    DiskCapacity     int
    Isp              string
    Bandwidth        int
    ImageType        string
    IpType           string
    GroupName        string
    client           Client
}

const (
    defaultURL = "http://api.speedycloud.cn/api/v1/products"
    defaultSSHUser = "root"
    defaultSSHPort = 22
    defaultActiveTimeout = 200
    defaultKeyPairName = "cloudos"
    defaultAvailabilityZone = "SPC-BJ-15-A"
    defaultAPISecret = "26685451e2a302886fb2a53958a787f65d3433731bee56d8f7fe96a714f677db"
    defaultAPIKey = "A44384F4AE3998DFEEAFD5E6CC6B1D56"
    defaultKeyFile = "/tmp/cloudos"
    defaultNetworkName = ""
    defaultCpuNumber = 2
    defaultMemory = 1024
    defaultDiskType = "Normal"
    defaultDiskCapacity = 20
    defaultISP = "Private"
    defaultBandwidth = 2
    defaultImage = "Ubuntu 14.04"
    defaultIpType = "inner"
    defaultUserDate = "/tmp/userdata"
    defaultGroupName = "cloudos"
)

func (d *Driver) GetCreateFlags() []mcnflag.Flag {
    return []mcnflag.Flag{
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_URL",
            Name:   "speedycloud-url",
            Usage:  "SpeedyCloud URL",
            Value:  defaultURL,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_API_KEY",
            Name:   "speedycloud-api-key",
            Usage:  "SpeedyCloud api key",
            Value:  defaultAPIKey,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_API_SECRET",
            Name:   "speedycloud-api-secret",
            Usage:  "SpeedyCloud password",
            Value:  defaultAPISecret,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_KEYPAIR_NAME",
            Name:   "speedycloud-keypair-name",
            Usage:  "SpeedCloud keypair to use to SSH to the instance",
            Value:  defaultKeyPairName,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_AVAILABILITY_ZONE",
            Name:   "speedycloud-availability-zone",
            Usage:  "SpeedyCloud zone where to create the instance ",
            Value:  defaultAvailabilityZone,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_PRIVATE_KEY_FILE",
            Name:   "speedycloud-private-key-file",
            Usage:  "Private keyfile to use for SSH (absolute path)",
            Value:  defaultKeyFile,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_USER_DATA_FILE",
            Name:   "speedycloud-user-data-file",
            Usage:  "File containing an SpeedyCloud userdata script",
            Value:  defaultUserDate,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_NETWORK_NAME",
            Name:   "speedycloud-net-name",
            Usage:  "SpeedyCloud network name the machine will be connected on",
            Value:  defaultNetworkName,
        },

        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_SSH_USER",
            Name:   "speedycloud-ssh-user",
            Usage:  "SpeedCloud SSH user",
            Value:  defaultSSHUser,
        },
        mcnflag.IntFlag{
            EnvVar: "SPEED_CLOUD_SSH_PORT",
            Name:   "speedycloud-ssh-port",
            Usage:  "SpeedyCloud SSH port",
            Value:  defaultSSHPort,
        },
        mcnflag.IntFlag{
            EnvVar: "SPEED_CLOUD_ACTIVE_TIMEOUT",
            Name:   "speedycloud-active-timeout",
            Usage:  "SpeedyCloud active timeout",
            Value:  defaultActiveTimeout,
        },
        mcnflag.IntFlag{
            EnvVar: "SPEED_CLOUD_CPU_NUMBER",
            Name:   "speedycloud-cpu-number",
            Usage:  "cpu number of the instance",
            Value:  defaultCpuNumber,
        },
        mcnflag.IntFlag{
            EnvVar: "SPEED_CLOUD_MEMORY",
            Name:   "speedycloud-memory",
            Usage:  "memory of the instance",
            Value:  defaultMemory,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_DISKTYPE",
            Name:   "speedycloud-disk-type",
            Usage:  "disk type of the instance",
            Value:  defaultDiskType,
        },
        mcnflag.IntFlag{
            EnvVar: "SPEED_CLOUD_DISKCAPACITY",
            Name:   "speedycloud-disk-capacity",
            Usage:  "disk capacity of the instance",
            Value:  defaultDiskCapacity,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_ISP",
            Name:   "speedycloud-isp",
            Usage:  "isp type of the instance",
            Value:  defaultISP,
        },
        mcnflag.IntFlag{
            EnvVar: "SPEED_CLOUD_BANDWIDTH",
            Name:   "speedycloud-bandwidth",
            Usage:  "bandwidth type of the instance",
            Value:  defaultBandwidth,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_IMAGE_TYPE",
            Name:   "speedycloud-image-type",
            Usage:  "imgage type of the instance",
            Value:  defaultImage,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_IP_TYPE",
            Name:   "speedycloud-ip-type",
            Usage:  "ip type of the instance, cloud be outer or inner",
            Value:  defaultIpType,
        },
        mcnflag.StringFlag{
            EnvVar: "SPEED_CLOUD_GROUP_NAME",
            Name:   "speedycloud-group-name",
            Usage:  "the group of the instance",
            Value:  defaultGroupName,
        },
    }
}

func NewDriver(hostName, storePath string) drivers.Driver {
    return NewDerivedDriver(hostName, storePath)
}

func NewDerivedDriver(hostName, storePath string) *Driver {
    return &Driver{
        client:        &GenericClient{},
        ActiveTimeout: defaultActiveTimeout,
        BaseDriver: &drivers.BaseDriver{
            SSHUser:     defaultSSHUser,
            SSHPort:     defaultSSHPort,
            MachineName: hostName,
            StorePath:   storePath,
        },
    }
}

func (d *Driver) GetSSHHostname() (string, error) {
    return d.GetIP()
}

func (d *Driver) SetClient(client Client) {
    d.client = client
}

// DriverName returns the name of the driver
func (d *Driver) DriverName() string {
    return "speedycloud"
}

func (d *Driver) SetConfigFromFlags(flags drivers.DriverOptions) error {
    d.SpeedCloudUrl = flags.String("speedycloud-url")
    d.ApiKey = flags.String("speedycloud-api-key")
    d.ApiSecret = flags.String("speedycloud-api-secret")
    d.KeyPairName = flags.String("speedycloud-keypair-name")
    d.AvailabilityZone = flags.String("speedycloud-availability-zone")
    d.PrivateKeyFile = flags.String("speedycloud-private-key-file")
    d.NetworkName = flags.String("speedycloud-net-name")
    d.SSHUser = flags.String("speedycloud-ssh-user")
    d.SSHPort = flags.Int("speedycloud-ssh-port")
    d.ActiveTimeout = flags.Int("speedycloud-active-timeout")
    d.CpuNumber = flags.Int("speedycloud-cpu-number")
    d.Memory = flags.Int("speedycloud-memory")
    d.DiskType = flags.String("speedycloud-disk-type")
    d.DiskCapacity = flags.Int("speedycloud-disk-capacity")
    d.Isp = flags.String("speedycloud-isp")
    d.Bandwidth = flags.Int("speedycloud-bandwidth")
    d.ImageType = flags.String("speedycloud-image-type")
    d.IpType = flags.String("speedycloud-ip-type")
    d.GroupName = flags.String("speedycloud-group-name")
    if flags.String("speedycloud-user-data-file") != "" {
        userData, err := ioutil.ReadFile(flags.String("speedycloud-user-data-file"))
        if err == nil {
            d.UserData = string(userData[:])
        } else {
            return err
        }
    }

    d.SetSwarmConfigFromFlags(flags)

    return d.checkConfig()
}

func (d *Driver) GetURL() (string, error) {
    if err := drivers.MustBeRunning(d); err != nil {
        return "", err
    }

    ip, err := d.GetIP()
    if err != nil {
        return "", err
    }
    if ip == "" {
        return "", nil
    }

    return fmt.Sprintf("tcp://%s", net.JoinHostPort(ip, "2376")), nil
}

func (d *Driver) GetIP() (string, error) {
    if d.IPAddress != "" {
        return d.IPAddress, nil
    }

    log.Debug("Looking for the IP address...", map[string]string{"MachineId": d.MachineId})

    if err := d.initCompute(); err != nil {
        return "", err
    }

    // Looking for the IP address in a retry loop to deal with OpenStack latency
    for retryCount := 0; retryCount < 200; retryCount++ {
        addresses, err := d.client.GetInstanceIPAddresses(d)
        if err != nil {
            return "", err
        }
        if d.IpType == "inner" {
            for _, a := range addresses {
                if match, _:= regexp.MatchString(`192.*|172.*|10.*`, a); match{
                    log.Debug("the address is %s", a)
                    return a, nil
                }
            }
        } else {
            for _, a := range addresses {
                if match, _:= regexp.MatchString(`192.*|172.*|10.*`, a); !match{
                    log.Info("the address is %s", a)
                    return a, nil
                }

            }
        }

        time.Sleep(2 * time.Second)

    }
    return "", fmt.Errorf("No IP found for the machine")
}

func (d *Driver) GetState() (state.State, error) {
    log.Debug("Get status for OpenStack instance...", map[string]string{"MachineId": d.MachineId})
    if err := d.initCompute(); err != nil {
        return state.None, err
    }

    s, err := d.client.GetInstanceState(d)
    if err != nil {
        return state.None, err
    }

    log.Debug("State for OpenStack instance", map[string]string{
        "MachineId": d.MachineId,
        "State":     s,
    })

    switch s {
    case "Running":
        return state.Running, nil
    case "Suspended":
        return state.Saved, nil
    case "Stopped":
        return state.Stopped, nil
    case "Provisioning":
        return state.Starting, nil
    case "ERROR":
        return state.Error, nil
    }
    return state.None, nil
}

func (d *Driver) Create() error {
    if err := d.resolveIds(); err != nil {
        return err
    }
    if d.KeyPairName != "" {
        if err := d.loadSSHKey(); err != nil {
            return err
        }
    } else {
        d.KeyPairName = fmt.Sprintf("%s-%s", d.MachineName, mcnutils.GenerateRandomID())
        if err := d.createSSHKey(); err != nil {
            return err
        }
    }
    if err := d.createMachine(); err != nil {
        return err
    }
    if err := d.waitForInstanceActive(); err != nil {
        return err
    }
    if err := d.lookForIPAddress(); err != nil {
        return err
    }

    if err := d.updateInstance(); err != nil {
        return err
    }
    return nil
}

func (d *Driver) Start() error {
    if err := d.initCompute(); err != nil {
        return err
    }

    return d.client.StartInstance(d)
}

func (d *Driver) Stop() error {
    if err := d.initCompute(); err != nil {
        return err
    }

    return d.client.StopInstance(d)
}

func (d *Driver) Restart() error {
    if err := d.initCompute(); err != nil {
        return err
    }

    return d.client.RestartInstance(d)
}

func (d *Driver) Kill() error {

    return d.Stop()
}

func (d *Driver) Remove() error {
    log.Debug("deleting instance...", map[string]string{"MachineId": d.MachineId})
    log.Info("Deleting speedycloud instance...")
    if err := d.initCompute(); err != nil {
        return err
    }
    if status, _ := d.GetState(); status != state.Stopped {
        d.Stop()
        d.waitForInstanceStopped()
    }
    if err := d.client.DeleteInstance(d); err != nil {
        return err
    }
    //log.Debug("deleting key pair...", map[string]string{"Name": d.KeyPairName})
    // TODO (fsoppelsa) maybe we want to check this, in case of shared keypairs, before removal
    //if err := d.client.DeleteKeyPair(d, d.KeyPairName); err != nil {
    //	return err
    //}
    return nil
}

const (
    errorMandatoryEnvOrOption string = "%s must be specified either using the environment variable %s or the CLI option %s"
    //errorMandatoryOption string = "%s must be specified using the CLI option %s"
    //errorExclusiveOptions string = "Either %s or %s must be specified, not both"
    //errorBothOptions string = "Both %s and %s must be specified"
    //errorMandatoryTenantNameOrID string = "Tenant id or name must be provided either using one of the environment variables OS_TENANT_ID and OS_TENANT_NAME or one of the CLI options --speedycloud-tenant-id and --speedycloud-tenant-name"
    //errorWrongEndpointType string = "Endpoint type must be 'publicURL', 'adminURL' or 'internalURL'"
    //errorUnknownFlavorName string = "Unable to find flavor named %s"
    //errorUnknownImageName string = "Unable to find image named %s"
    //errorUnknownNetworkName string = "Unable to find network named %s"
    //errorUnknownTenantName string = "Unable to find tenant named %s"
)

func (d *Driver) checkConfig() error {

    if d.SpeedCloudUrl == "" {
        return fmt.Errorf(errorMandatoryEnvOrOption, "Speedycloud url", "SPEED_CLOUD_URL", "--speedycloud-url")
    }

    if d.ApiKey == "" {
        return fmt.Errorf(errorMandatoryEnvOrOption, "Api key", "SPEED_CLOUD_API_KEY", "speedycloud-api-key")
    }
    if d.ApiSecret == "" {
        return fmt.Errorf(errorMandatoryEnvOrOption, "Api secret", "SPEED_CLOUD_API_SECRET", "speedycloud-api-secret")
    }
    //if d.KeyPairName == "" {
    //    return fmt.Errorf(errorMandatoryEnvOrOption, "Key Pair Name", "SPEED_CLOUD_KEYPAIR_NAME", "speedycloud-keypair-name")
    //}
    if d.AvailabilityZone == "" {
        return fmt.Errorf(errorMandatoryEnvOrOption, "AZ", "SPEED_CLOUD_AVAILABILITY_ZONE", "speedycloud-availability-zone")
    }
    //if d.PrivateKeyFile == "" {
    //    return fmt.Errorf(errorMandatoryEnvOrOption, "Private Key File", "SPEED_CLOUD_PRIVATE_KEY_FILE", "speedycloud-private-key-file")
    //}
    if d.SSHUser == "" {
        return fmt.Errorf(errorMandatoryEnvOrOption, "Ssh User ", "SPEED_CLOUD_SSH_USER", "speedycloud-ssh-user")
    }
    if string(d.SSHPort) == "" {
        return fmt.Errorf(errorMandatoryEnvOrOption, "Ssh Port", "SPEED_CLOUD_SSH_PORT", "speedycloud-ssh-port")
    }
    if string(d.CpuNumber) == "" {
        return fmt.Errorf(errorMandatoryEnvOrOption, "Cpu Number", "SPEED_CLOUD_CPU_NUMBER", "speedycloud-cpu-number")
    }
    if string(d.Memory) == "" {
        return fmt.Errorf(errorMandatoryEnvOrOption, "Memory", "SPEED_CLOUD_MEMORY", "speedycloud-memory")
    }
    if d.ImageType == "" {
        return fmt.Errorf(errorMandatoryEnvOrOption, "Key Pair Name", "SPEED_CLOUD_IMAGE_TYPE", "speedycloud-image-type")
    }

    return nil
}

func (d *Driver) resolveIds() error {

    return nil
}

func (d *Driver) initCompute() error {
    if err := d.client.InitProviderClient(d); err != nil {
        return err
    }
    if err := d.client.InitComputeClient(d); err != nil {
        return err
    }
    return nil
}

func (d *Driver) initNetwork() error {
    if err := d.client.InitProviderClient(d); err != nil {
        return err
    }
    if err := d.client.InitNetworkClient(d); err != nil {
        return err
    }
    return nil
}

func (d *Driver) loadSSHKey() error {
    log.Debug("Loading Key Pair", d.KeyPairName)
    if err := d.initCompute(); err != nil {
        return err
    }
    log.Debug("Loading Private Key from", d.PrivateKeyFile)
    privateKey, err := ioutil.ReadFile(d.PrivateKeyFile)
    if err != nil {
        return err
    }
    publicKey, err := d.client.GetPublicKey(d.KeyPairName)
    if err != nil {
        return err
    }
    if err := ioutil.WriteFile(d.privateSSHKeyPath(), privateKey, 0600); err != nil {
        return err
    }
    if err := ioutil.WriteFile(d.publicSSHKeyPath(), publicKey, 0600); err != nil {
        return err
    }

    return nil
}

func (d *Driver) createSSHKey() error {
    sanitizeKeyPairName(&d.KeyPairName)
    log.Debug("Creating Key Pair...", map[string]string{"Name": d.KeyPairName})
    if err := ssh.GenerateSSHKey(d.GetSSHKeyPath()); err != nil {
        return err
    }
    publicKey, err := ioutil.ReadFile(d.publicSSHKeyPath())
    if err != nil {
        return err
    }

    if err := d.initCompute(); err != nil {
        return err
    }
    if err := d.client.CreateKeyPair(d, d.KeyPairName, string(publicKey)); err != nil {
        return err
    }
    return nil
}

func (d *Driver) createMachine() error {
    log.Debug("Creating SpeedyCloud instance...", map[string]string{
        "Cpu": string(d.CpuNumber),
        "Memory": string(d.Memory),
        "diskCapacity": string(d.DiskCapacity),
        "ImageId":  d.ImageType,
    })

    if err := d.initCompute(); err != nil {
        return err
    }
    instanceID, err := d.client.CreateInstance(d)
    if err != nil {
        return err
    }
    d.MachineId = instanceID
    return nil
}

//func (d *Driver) assignFloatingIP() error {
//    var err error
//
//    if d.ComputeNetwork {
//        err = d.initCompute()
//    } else {
//        err = d.initNetwork()
//    }
//
//    if err != nil {
//        return err
//    }
//
//    ips, err := d.client.GetFloatingIPs(d)
//    if err != nil {
//        return err
//    }
//
//    var floatingIP *FloatingIP
//
//    log.Debugf("Looking for an available floating IP", map[string]string{
//        "MachineId": d.MachineId,
//        "Pool":      d.FloatingIpPool,
//    })
//
//    for _, ip := range ips {
//        if ip.PortId == "" {
//            log.Debug("Available floating IP found", map[string]string{
//                "MachineId": d.MachineId,
//                "IP":        ip.Ip,
//            })
//            floatingIP = &ip
//            break
//        }
//    }
//
//    if floatingIP == nil {
//        floatingIP = &FloatingIP{}
//        log.Debug("No available floating IP found. Allocating a new one...", map[string]string{"MachineId": d.MachineId})
//    } else {
//        log.Debug("Assigning floating IP to the instance", map[string]string{"MachineId": d.MachineId})
//    }
//
//    if err := d.client.AssignFloatingIP(d, floatingIP); err != nil {
//        return err
//    }
//    d.IPAddress = floatingIP.Ip
//    return nil
//}

func (d *Driver) waitForInstanceActive() error {
    log.Debug("Waiting for the SpeedyCloud instance to be running...", map[string]string{"MachineId": d.MachineId})
    if err := d.client.WaitForInstanceStatus(d, "Running"); err != nil {
        return err
    }
    return nil
}

func (d *Driver) updateInstance() error {
    log.Debug("update SpeedyCloud instance group and alias", map[string]string{"MachineId": d.MachineId})
    if err := d.client.UpdateInstanceAlias(d); err != nil {
        return err
    }
    if err := d.client.UpdateInstanceGroup(d); err != nil {
        return err
    }
    return nil
}

func (d *Driver) waitForInstanceStopped() error {
    log.Debug("Waiting for the SpeedyCloud instance to be stopped...", map[string]string{"MachineId": d.MachineId})
    if err := d.client.WaitForInstanceStatus(d, "Stopped"); err != nil {
        return err
    }
    return nil
}

func (d *Driver) lookForIPAddress() error {
    ip, err := d.GetIP()
    if err != nil {
        return err
    }
    d.IPAddress = ip
    log.Debug("IP address found", map[string]string{
        "IP":        ip,
        "MachineId": d.MachineId,
    })
    return nil
}

func (d *Driver) privateSSHKeyPath() string {
    return d.GetSSHKeyPath()
}

func (d *Driver) publicSSHKeyPath() string {
    return d.GetSSHKeyPath() + ".pub"
}

func sanitizeKeyPairName(s *string) {
    *s = strings.Replace(*s, ".", "_", -1)
}
